bandwhich: add shell completions and man page, add `meta.changelog`

emilylange opened the pull request:
## Description of changes

Pre v0.22.0, shell completions and the man page were static and heavily outdated.

v0.22.0 then removed those entirely.

v0.22.2 then re-added shell completions and the man page, but this time natively generated on build time (assuming $BANDWHICH_GEN_DIR is set).

Technically, shell completions for PowerShell and Elvish are generated as well, but we don't install those, since our installShellCompletions does not support those shells.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

emilylange committed:
bandwhich: add shell completions and man page

Pre v0.22.0, shell completions and the man page were static and heavily
outdated.

v0.22.0 then removed those entirely.

v0.22.2 then re-added shell completions and the man page, but this time
natively generated on build time (assuming $BANDWHICH_GEN_DIR is set).

Technically, shell completions for PowerShell and Elvish are generated
as well, but we don't install those, since our installShellCompletions
does not support those shells.

SuperSandro2000 commented on pkgs/tools/networking/bandwhich/default.nix:
```diff
@@ -23,11 +23,25 @@ rustPlatform.buildRustPackage rec {
+  postInstall = ''
+    installManPage $BANDWHICH_GEN_DIR/bandwhich.1
+
+    installShellCompletion $BANDWHICH_GEN_DIR/bandwhich.{bash,fish} \
```
Does this work like it should? I would expect the below to work reliable.


```suggestion
    installShellCompletion --cmd bandwhich \
      --bash $BANDWHICH_GEN_DIR/bandwhich.bash \
      --fish $BANDWHICH_GEN_DIR/bandwhich.fish \
```

emilylange replied:
`installShellCompletion` has a bunch of auto-detection logic[^1].

This essentially boils down to:
- `--cmd` isn't needed, because the completions already contain the command (e.g. `bandwhich` in `bandwhich.bash`)
- `bandwhich.bash` and `bandwhich.fish` have proper file extensions (`.bash` and `.fish`) from which the shell is automatically inferred[^2]
- `_bandwhich` for zsh is the only completion that needs special care, unless one wants an ugly `installShellCompletion: warning: assuming path $arg is zsh; please specify with --zsh` warning  [^3]

But I can use your version/variant, if you prefer it that way. No hard feelings.

[^1]: https://github.com/NixOS/nixpkgs/blob/55ad4874cf42a9c1425840a9907a68234b74945a/pkgs/build-support/setup-hooks/install-shell-files.sh#L52-L97

[^2]: https://github.com/NixOS/nixpkgs/blob/55ad4874cf42a9c1425840a9907a68234b74945a/pkgs/build-support/setup-hooks/install-shell-files.sh#L162-L167

[^3]: https://github.com/NixOS/nixpkgs/blob/55ad4874cf42a9c1425840a9907a68234b74945a/pkgs/build-support/setup-hooks/install-shell-files.sh#L168-L172

emilylange replied:
@SuperSandro2000 and @figsoda are you fine with me marking this as resolved?

SuperSandro2000 replied:
fine for me

emilylange committed:
bandwhich: add `meta.changelog`

bandwhich has two different changelogs:
1. the in-tree CHANGELOG.md
2. git tags/GitHub Releases

Unfortunately, the CHANGELOG.md contains fewer information from time to
time. One example would be
https://github.com/imsnif/bandwhich/blob/v0.22.0/CHANGELOG.md vs
https://github.com/imsnif/bandwhich/releases/tag/v0.22.0

In this case, the GitHub Release contains a lot more noise due to
various chore(deps) entries.

This is due to
https://github.com/imsnif/bandwhich/blob/b5e48ba9bb466c5e1192755b0235b6b43d1d4966/.github/workflows/require-changelog-for-PRs.yml#L10-L11

I'd argue, the CHANGELOG.md should be preferred here, but in the end
this isn't my package and not my call.

dywedir commented:
Thank you!
